driver/src/orderbook/streamed/order.rs

```diff
@@ -1,19 +1,106 @@
 use super::*;
+use crate::contracts::stablex_auction_element;
+use crate::models::Order as ModelOrder;
+use anyhow::Result;
 use serde::{Deserialize, Serialize};
 
+/// Change from a potential solution that might still get replaced by a better solution in the same
+/// batch.
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+struct PendingUsedAmount {
+    batch_id: BatchId,
+    amount: u128,
+}
+
 #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
 pub struct Order {
     pub buy_token: TokenId,
     pub sell_token: TokenId,
-    pub valid_from: BatchId,
     pub valid_until: BatchId,
-    pub price_numerator: u128,
-    pub price_denominator: u128,
-    pub used_amount: u128,
+    pub valid_from: BatchId,
+    price_numerator: u128,
+    price_denominator: u128,
+    // Invariant: used_amount + pending_used_amount.amount <= price_denominator
+    used_amount: u128,
+    pending_used_amount: PendingUsedAmount,
```

Comment:
I don't understand why this is necessary, doesn't `TradeReversion` contain this data? I also thought that events were going to get applied in a way that only the relevant events (that is up until the batch that we are requesting the orderbook for) so this shouldn't be necessary for ensuring that the trade is not included in the orderbook right?

Reply:
I believe the reason is that for some orderbooks you don't want to take the pending proceeds into account (namely the one for which the proceed is pending in case we want to resolve it after a solution has already been submitted), but for others you do (e.g. the next batch for which we optimistically assume the current solution will stay the best one).

Thus we need to have a way even without a new event to conditionally deduct or incorporate pending proceeds.

Reply:
> in case we want to resolve the batch after another solver has already submitted a solution [...] but for other orderbooks you want to include them

I was under the impression that we were going to keep all the events in memory and then apply only the needed events to build the orderbook. In that case aren't the events for previous solution just not going to be applied? If we want to query the open orderbook I would imagine that just more events would be applied (like the current winning solution).

Reply:
The way I understood what we decided and implemented it is that the "event manager" is completely agnostic about how events work. It applies all of them because it doesn't know which events are relevant. This is why this code needs to handle this as Felix said.

Reply:
I see the comment below, I guess I slightly misunderstood what we ended up deciding. This is indeed needed then!.

Reply:
> what we decided is that the "event manager" is completely agnostic about how events work

Yes, I had understood that it just meant that it would not inspect data to decide if an event should be included or not.